# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1499372738 14400
#      Thu Jul 06 16:25:38 2017 -0400
# Node ID 38188d1aaf3f4b013e904cc7be07289515868d4f
# Parent  884e18e82e8f9fca94d188d425fa7134bb2abdcd
ensure that temporary file used by gl2ps is closed (bug #50696)

* gl2ps-print.cc (gl2ps_renderer:draw): Use unwind-protect to ensure
that temporary file is closed.

diff --git a/libinterp/corefcn/gl2ps-print.cc b/libinterp/corefcn/gl2ps-print.cc
--- a/libinterp/corefcn/gl2ps-print.cc
+++ b/libinterp/corefcn/gl2ps-print.cc
@@ -43,16 +43,30 @@ along with Octave; see the file COPYING.
 #include "unwind-prot.h"
 
 #include "gl-render.h"
 #include "oct-opengl.h"
 #include "sighandlers.h"
 #include "sysdep.h"
 #include "text-renderer.h"
 
+static void
+safe_pclose (FILE *f)
+{
+  if (f)
+    octave_pclose (f);
+}
+
+static void
+safe_fclose (FILE *f)
+{
+  if (f)
+    std::fclose (f);
+}
+
 namespace octave
 {
   class
   OCTINTERP_API
   gl2ps_renderer : public opengl_renderer
   {
   public:
 
@@ -233,16 +247,18 @@ namespace octave
           gl2ps_sort = GL2PS_NO_SORT;
 
         // Use a temporary file in case an overflow happens
         FILE *tmpf = octave_tmpfile_wrapper ();
 
         if (! tmpf)
           error ("gl2ps_renderer::draw: couldn't open temporary file for printing");
 
+        frame.add_fcn (safe_fclose, tmpf);
+
         // Reset buffsize, unless this is 2nd pass of a texstandalone print.
         if (term.find ("tex") == std::string::npos)
           buffsize = 2*1024*1024;
         else
           buffsize /= 2;
 
         buffer_overflow = true;
 
@@ -807,30 +823,16 @@ namespace octave
     std::string str = props.get_string ().string_vector_value ().join ("\n");
 
     render_text (str, pos(0), pos(1), pos.numel () > 2 ? pos(2) : 0.0,
                  halign, valign, props.get_rotation ());
   }
 
 }
 
-static void
-safe_pclose (FILE *f)
-{
-  if (f)
-    octave_pclose (f);
-}
-
-static void
-safe_fclose (FILE *f)
-{
-  if (f)
-    std::fclose (f);
-}
-
 #endif
 
 namespace octave
 {
 
   // If the name of the stream begins with '|', open a pipe to the command
   // named by the rest of the string.  Otherwise, write to the named file.
 
